File: Programacao-III/java-sistema-eleitoral/codigo/src/Pessoa.java
import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;

public class Pessoa {
  private String nome;
  private char sexo;
  private LocalDate dataNascimento;
  
	public Pessoa(String nome, char sexo, LocalDate dataNascimento) {
		this.nome = nome;
		this.sexo = sexo;
		this.dataNascimento = dataNascimento;
	}

	public Pessoa(String nome, char sexo, String dataNascimento) {
		this.nome = nome;
		this.sexo = sexo;
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
		this.dataNascimento = LocalDate.parse(dataNascimento, formatter);
	}
	
	
	/**
	 * Retorna o nome 
	 * @return String nome 
	 */
	public String getNome() {
		return nome;
	}
	
	/**
	 * Altera o nome 
	 * @param nome String nome 
	 */
	public void setNome(String nome) {
		this.nome = nome;
	}
	
	/**
	 * Retorna o sexo
	 * @return char sexo: F ou M 
	 */
	public char getSexo() {
		return sexo;
	}
	
	/**
	 * Altera o sexo
	 * @param sexo char sexo: F ou M 
	 */
	public void setSexo(char sexo) {
		this.sexo = sexo;
	}
	
	/**
	 * Retorna a data de nascimento 
	 * @return LocalDate data de nascimento 
	 */
	public LocalDate getDataNascimento() {
		return dataNascimento;
	}
	
	/**
	 * Altera a data de nascimento 
	 * @param dataNascimento LocalDate data de nascimento 
	 */
	public void setDataNascimento(LocalDate dataNascimento) {
		this.dataNascimento = dataNascimento;
	}
	
	/**
	 * Altera a data de nascimento 
	 * @param dataNascimento String data de nascimento
	 */
	public void setDataNascimento(String dataNascimento) {
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
		this.dataNascimento = LocalDate.parse(dataNascimento, formatter);
	}
	
	/**
	 * Cálcula a idade da pessoa 
	 * @return int idade da pessoa 
	 * ou 0 caso data de nascimento ou a data de referência sejam nulos
	 */
	public int calculaIdade() {
        if ((dataNascimento != null) && (Eleicao.getDataAtual() != null)) {
            return Period.between(dataNascimento, Eleicao.getDataAtual()).getYears();
        } else {
            return 0;
        }
    }
	/**
	 * Compara idade entre duas pessoas
	 * @param o2 Pessoa para comparar
	 * @return int 
	 * se o2 é mais velho valor é negativo
	 * se o2 é mais novo valor é positivo
	 */
	public int comparaIdade(Pessoa o2) {
		return (this.calculaIdade() - o2.calculaIdade());
	}
}

File: Programacao-III/cpp-sistema-eleitoral/Pessoa.cpp
#include <iostream>
#include "Pessoa.h"

using namespace std;

Pessoa::Pessoa() {}

Pessoa::Pessoa(const string &nome, const string &sexo, const string &dataNascimento)
{
  this->nome = nome;
  this->sexo = sexo;
  this->dataNascimento = parseDate(dataNascimento, DATE_FORMAT_PT_BR_SHORT);
}

Pessoa::Pessoa(const Pessoa &o)
{
  this->nome = o.nome;
  this->sexo = o.sexo;
  this->dataNascimento = o.dataNascimento;
}

Pessoa &Pessoa::operator=(const Pessoa &o)
{
  this->nome = o.nome;
  this->sexo = o.sexo;
  this->dataNascimento = o.dataNascimento;
  return *this;
}

const string &Pessoa::getNome() const
{
  return this->nome;
}

void Pessoa::setNome(const string &nome)
{
  this->nome = nome;
}

const string &Pessoa::getSexo() const
{
  return this->sexo;
}

void Pessoa::setSexo(const string &sexo)
{
  this->sexo = sexo;
}

const time_t &Pessoa::getDataNascimento() const
{
  return this->dataNascimento;
}

void Pessoa::setDataNascimento(const time_t &dataNascimento)
{
  this->dataNascimento = dataNascimento;
}

int Pessoa::calculaIdade(const time_t &dataAtual) const {
  tm *tm = localtime(&dataAtual);
  int dayda = tm->tm_mday;
  int monthda = tm->tm_mon + 1;
  int yearda = tm->tm_year + 1900;
  tm = localtime(&dataNascimento);
  int daydn = tm->tm_mday;
  int monthdn = tm->tm_mon + 1;
  int yeardn = tm->tm_year + 1900;
  int idade = yearda - yeardn;
  if ((monthdn == monthda && daydn > dayda) || monthdn > monthda) {
    return idade - 1;
  }
  return idade;
}

int Pessoa::comparaIdade(Pessoa *o2)
{
  if (difftime(this->dataNascimento, o2->dataNascimento) < 0)
    return -1;
  else if (difftime(this->dataNascimento, o2->dataNascimento) == 0)
    return 0;
  return 1;
}

Pessoa::~Pessoa() {}